File: Leetcode/maximum_subarray.py
# https://leetcode.com/problems/maximum-subarray/
'''
    Author - Subhajit Das
    University of Engineering and Management, Kolkata
    10/07/2019
'''


import sys


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        dp0, dp1 = 0, None
        m = -(sys.maxsize)+1

        for i in range(n):
            dp1 = max(dp0 + nums[i], nums[i])
            if(m < dp1):
                m = dp1
            dp0 = dp1
        return m

File: Leetcode/MaximumSubarray.java
// https://leetcode.com/problems/maximum-subarray/
/*
    Author @Subhajit Das (sdsubhajitdas.github.io)
    SWE @Turbot HQ India PVT Ltd.
    16/04/2022
*/

import java.util.*;

class Solution {
  public int maxSubArray(int[] nums) {
    int sum=0, maxSum = Integer.MIN_VALUE;

    for(int i =0; i<nums.length; i++) {
      sum += nums[i];

      maxSum = Integer.max(maxSum,sum);
      if(sum < 0) 
        sum = 0;

    }

    return maxSum;
  }
}